README
BSDNT is a BSD licensed bignum library, suitable for use by languages which
need to provide bignum support and for other projects requiring relatively
fast bignums.

The primary goal of the project is to have clean, comprehensible, 
well-documented code, whilst retaining reasonable performance. It is not
a goal to insanely optimise where that would make the code less readable
and maintainable.

The project will be developed in numerous stages:

i) A C only phase, implementing nothing but basecase algorithms, which will
serve to map out the interface. This code will eventually be used as test 
code and fallback C code for architectures without assembly support.

ii) Addition of algorithms to make the library efficient up to a few hundred 
limbs and inline assembly code for x86 and x86_64.

iii) Addition of asymptotically fast algorithms for huge operands.

iv) Addition of assembly code for other platforms.

At least the first phase of the project will be completed as a github social
coding project, with a blog every few days which interested parties can 
follow along with their own github fork in the language of their choice.

It's expected that step 1 will take a few weeks. Step 2 will take 6 months to
a year.

----

To build bsdnt and run the test code

$ cd bsdnt
$ ./configure
$ make check

----

See the LICENSE file for legal info and warranty information.

Enjoy!!



CHANGES
04-Sep-2010:
	First commit - CHANGES and LICENSE file.

05-Sep-2010:
 	v0.1 - nn_add function, random functions, nn_equal_m, nn_equal, 
	basic types.

05-Sep-2010:
        v0.2 - nn_sub_m, nn_lsh, nn_rsh.

06-Sep-2010:
        v0.3 - nn_copy, nn_zero, nn_normalise, nn_mul1.

06-Sep-2010:
        v0.4 - nn_not, nn_neg, nn_add1, nn_sub1.

07-Sep-2010:
        v0.5 - nn_add, nn_sub, nn_cmp_m, nn_cmp.

07-Sep-2010:
	v0.6 - nn_addmul1, nn_submul1.

08-Sep-2010:
	v0.7 - nn_divrem1_simple.

08-Sep-2010:
	v0.8 - nn_divrem1_preinv.

09-Sep-2010:
	v0.9 - nn_divrem_hensel1_preinv.
    t-nn_linear.o: t-nn_linear.c

09-Sep-2010:
	v0.10 - nn_mod1_preinv

10-Sep-2010:
	v0.11 - generics.

12-Sep-2010:
	v0.12 - nn_mul_classical.

19-Sep-2010:
        asm2 - assembly code for nn_add/sub_mc, configure.

17-Sep-2010:
	v0.13 - Added nn_muladd1 and nn_muladd_classical.

19-Sep-2010:
	v0.14 - Added nn_divrem_classical.

20-Sep-2010:
	v0.15 - Added nn_divapprox_classical.

25-Sep-2010:
	v0.16 - Added nn_mullow_classical and nn_mulhigh_classical.

26-Sep-2010:
	v0.17 - Added nn_div_hensel_preinv.

01-Oct-2010:
	v0.18 - Added printx_word, nn_printx, nn_printx_short, test diagnostics.

02-Oct-2010:
	v0.19 - Added ASSERT macro.

06-Oct-2010:
	v0.20 - Redzones.

14-Oct-2010:
	v0.21 - Added KISS, SKISS, MT PRNGs.

15-Oct-2010:
	v0.22 - Added Windows support and new configure system.

28-Oct-2010:
	v0.23 - sha1 and mt, kiss and skiss tests.

29-Oct-2010:
	v0.24 - nn_bit_test/clear/set, nn_test_random/1/2.

31-Oct-2010:
	v0.25 - test_randword/1/2.

10-Nov-2010:
	v0.26 - bsdnt_printf.

24-Oct-2011:
	v0.27 - stopped quadratic division functions rereading inputs

11-Mar-2013:
	v0.28 - removed MSVC build, cleaned up configure removed _r/s_ 
                functions, removed muladd functions, switched to simpler
                division code.

12-Mar-2013:
	v0.29 - switched to gas assembly format.

12-Mar-2013:
	v0.30 - Added TMP_INIT/START/ALLOC/ALLOC_BYTES/END.

13-Mar-2013:
	v0.31 - Added nn_mul_kara, nn_mul_toom33, nn_mul_toom32, nn_mul and
                nn_mul_m.

23-Mar-2013:
	v0.32 - Added nn_divapprox_divconquer, nn_divrem_divconquer.

28-Mar-2013:
	v0.33 - Added nn_mulmid_kara and nn_mullow_kara.

15-Apr-2013:
	v0.34 - Added nn_div_divconquer, nn_mullow, nn_mullow_m and
                nn_mullow_kara_m, nn_divrem, nn_div.



sha1.c
/*
 ---------------------------------------------------------------------------
 Copyright (c) 1998-2008, Brian Gladman, Worcester, UK. All rights reserved.

 LICENSE TERMS

 The redistribution and use of this software (with or without changes)
 is allowed without the payment of fees or royalties provided that:

  1. source code distributions include the above copyright notice, this
     list of conditions and the following disclaimer;

  2. binary distributions include the above copyright notice, this list
     of conditions and the following disclaimer in their documentation;

  3. the name of the copyright holder is not used to endorse products
     built using this software without specific written permission.

 DISCLAIMER

 This software is provided 'as is' with no explicit or implied warranties
 in respect of its properties, including, but not limited to, correctness
 and/or fitness for purpose.
 ---------------------------------------------------------------------------
 Issue Date: 20/12/2007
*/

#include <string.h>     /* for memcpy() etc.        */

#include "sha1.h"
#include "config.h"

#define rotl32(x,n)   (((x) << n) | ((x) >> (32 - n)))
#define rotr32(x,n)   (((x) >> n) | ((x) << (32 - n)))

#if !defined(bswap_32)
#define bswap_32(x) ((rotr32((x), 24) & 0x00ff00ff) | (rotr32((x), 8) & 0xff00ff00))
#endif

#if (PLATFORM_BYTE_ORDER == IS_LITTLE_ENDIAN)
#define bsw_32(p,n) \
    { int _i = (n); while (_i--) ((uint32_t *) p)[_i] = bswap_32(((uint32_t *) p)[_i]); }
#else
#define bsw_32(p,n)
#endif

#define SHA1_MASK   (SHA1_BLOCK_SIZE - 1)

#if 0

#define ch(x,y,z)       (((x) & (y)) ^ (~(x) & (z)))
#define parity(x,y,z)   ((x) ^ (y) ^ (z))
#define maj(x,y,z)      (((x) & (y)) ^ ((x) & (z)) ^ ((y) & (z)))

#else   /* Discovered by Rich Schroeppel and Colin Plumb   */

#define ch(x,y,z)       ((z) ^ ((x) & ((y) ^ (z))))
#define parity(x,y,z)   ((x) ^ (y) ^ (z))
#define maj(x,y,z)      (((x) & (y)) | ((z) & ((x) ^ (y))))

#endif

/* Compile 64 bytes of hash data into SHA1 context. Note    */
/* that this routine assumes that the byte order in the     */
/* ctx->wbuf[] at this point is in such an order that low   */
/* address bytes in the ORIGINAL byte stream will go in     */
/* this buffer to the high end of 32-bit words on BOTH big  */
/* and little endian systems                                */

#ifdef ARRAY
#define q(v,n)  v[n]
#else
#define q(v,n)  v##n
#endif

#ifdef SHA_1

#define one_cycle(v,a,b,c,d,e,f,k,h)            \
    q(v,e) += rotr32(q(v,a),27) +               \
              f(q(v,b),q(v,c),q(v,d)) + k + h;  \
    q(v,b)  = rotr32(q(v,b), 2)

#define five_cycle(v,f,k,i)                 \
    one_cycle(v, 0,1,2,3,4, f,k,hf(i  ));   \
    one_cycle(v, 4,0,1,2,3, f,k,hf(i+1));   \
    one_cycle(v, 3,4,0,1,2, f,k,hf(i+2));   \
    one_cycle(v, 2,3,4,0,1, f,k,hf(i+3));   \
    one_cycle(v, 1,2,3,4,0, f,k,hf(i+4))

void sha1_compile(sha1_ctx ctx[1])
{   
	 uint32_t * w = ctx->wbuf;

#ifdef ARRAY
    uint32_t v[5];
    memcpy(v, ctx->hash, sizeof(ctx->hash));
#else
    uint32_t v0, v1, v2, v3, v4;
    v0 = ctx->hash[0]; v1 = ctx->hash[1];
    v2 = ctx->hash[2]; v3 = ctx->hash[3];
    v4 = ctx->hash[4];
#endif

#define hf(i) w[i]

    five_cycle(v, ch, 0x5a827999,  0);
    five_cycle(v, ch, 0x5a827999,  5);
    five_cycle(v, ch, 0x5a827999, 10);
    one_cycle(v,0,1,2,3,4, ch, 0x5a827999, hf(15)); \

#undef  hf
#define hf(i) (w[(i) & 15] = rotl32(                    \
                 w[((i) + 13) & 15] ^ w[((i) + 8) & 15] \
               ^ w[((i) +  2) & 15] ^ w[(i) & 15], 1))

    one_cycle(v,4,0,1,2,3, ch, 0x5a827999, hf(16));
    one_cycle(v,3,4,0,1,2, ch, 0x5a827999, hf(17));
    one_cycle(v,2,3,4,0,1, ch, 0x5a827999, hf(18));
    one_cycle(v,1,2,3,4,0, ch, 0x5a827999, hf(19));

    five_cycle(v, parity, 0x6ed9eba1,  20);
    five_cycle(v, parity, 0x6ed9eba1,  25);
    five_cycle(v, parity, 0x6ed9eba1,  30);
    five_cycle(v, parity, 0x6ed9eba1,  35);

    five_cycle(v, maj, 0x8f1bbcdc,  40);
    five_cycle(v, maj, 0x8f1bbcdc,  45);
    five_cycle(v, maj, 0x8f1bbcdc,  50);
    five_cycle(v, maj, 0x8f1bbcdc,  55);

    five_cycle(v, parity, 0xca62c1d6,  60);
    five_cycle(v, parity, 0xca62c1d6,  65);
    five_cycle(v, parity, 0xca62c1d6,  70);
    five_cycle(v, parity, 0xca62c1d6,  75);

#ifdef ARRAY
    ctx->hash[0] += v[0]; ctx->hash[1] += v[1];
    ctx->hash[2] += v[2]; ctx->hash[3] += v[3];
    ctx->hash[4] += v[4];
#else
    ctx->hash[0] += v0; ctx->hash[1] += v1;
    ctx->hash[2] += v2; ctx->hash[3] += v3;
    ctx->hash[4] += v4;
#endif
}

void sha1_begin(sha1_ctx ctx[1])
{
    memset(ctx, 0, sizeof(sha1_ctx));
    ctx->hash[0] = 0x67452301;
    ctx->hash[1] = 0xefcdab89;
    ctx->hash[2] = 0x98badcfe;
    ctx->hash[3] = 0x10325476;
    ctx->hash[4] = 0xc3d2e1f0;
}

/* SHA1 hash data in an array of bytes into hash buffer and */
/* call the hash_compile function as required. For both the */
/* bit and byte orientated versions, the block length 'len' */
/* must not be greater than 2^32 - 1 bits (2^29 - 1 bytes)  */ 

void sha1_hash(const unsigned char data[], unsigned long len, sha1_ctx ctx[1])
{   
	uint32_t pos = (uint32_t) ((ctx->count[0] >> 3) & SHA1_MASK);
    const unsigned char *sp = data;
    unsigned char *w = (unsigned char *) ctx->wbuf;
#if SHA1_BITS == 1
    uint32_t ofs = (ctx->count[0] & 7);
#else
    len <<= 3;
#endif
    if ((ctx->count[0] += len) < len)
        ++(ctx->count[1]);
#if SHA1_BITS == 1
    if (ofs)                 /* if not on a byte boundary   */
    {
        if(ofs + len < 8)   /* if no added bytes are needed */
        {
            w[pos] |= (*sp >> ofs);
        } else                /* otherwise and add bytes    */
        {   
			unsigned char part = w[pos];

            while ((int) (ofs + (len -= 8)) >= 0)
            {
                w[pos++] = part | (*sp >> ofs);
                part = *sp++ << (8 - ofs);
                if (pos == SHA1_BLOCK_SIZE)
                {
                    bsw_32(w, SHA1_BLOCK_SIZE >> 2);
                    sha1_compile(ctx); pos = 0;
                }
            }

            w[pos] = part;
        }
    } else    /* data is byte aligned */
#endif
    {   
		uint32_t space = SHA1_BLOCK_SIZE - pos;

        while (len >= (space << 3))
        {
            memcpy(w + pos, sp, space);
            bsw_32(w, SHA1_BLOCK_SIZE >> 2);
            sha1_compile(ctx); 
            sp += space; len -= (space << 3); 
            space = SHA1_BLOCK_SIZE; pos = 0;
        }

        memcpy(w + pos, sp, (len + 7 * SHA1_BITS) >> 3);
    }
}

/* SHA1 final padding and digest calculation  */

void sha1_end(unsigned char hval[], sha1_ctx ctx[1])
{   
	uint32_t i = (uint32_t) ((ctx->count[0] >> 3) & SHA1_MASK), m1;

    /* put bytes in the buffer in an order in which references to   */
    /* 32-bit words will put bytes with lower addresses into the    */
    /* top of 32 bit words on BOTH big and little endian machines   */
    bsw_32(ctx->wbuf, (i + 3 + SHA1_BITS) >> 2);

    /* we now need to mask valid bytes and add the padding which is */
    /* a single 1 bit and as many zero bits as necessary. Note that */
    /* we can always add the first padding byte here because the    */
    /* buffer always has at least one empty slot                    */
    m1 = (unsigned char) 0x80 >> (ctx->count[0] & 7);
    ctx->wbuf[i >> 2] &= ((0xffffff00 | (~m1 + 1)) << 8 * (~i & 3));
    ctx->wbuf[i >> 2] |= (m1 << 8 * (~i & 3));

    /* we need 9 or more empty positions, one for the padding byte  */
    /* (above) and eight for the length count. If there is not      */
    /* enough space, pad and empty the buffer                       */
    if(i > SHA1_BLOCK_SIZE - 9)
    {
        if(i < 60) ctx->wbuf[15] = 0;
        sha1_compile(ctx);
        i = 0;
    } else  /* compute a word index for the empty buffer positions  */
        i = (i >> 2) + 1;

    while (i < 14) /* and zero pad all but last two positions       */
        ctx->wbuf[i++] = 0;

    /* the following 32-bit length fields are assembled in the      */
    /* wrong byte order on little endian machines but this is       */
    /* corrected later since they are only ever used as 32-bit      */
    /* word values.                                                 */
    ctx->wbuf[14] = ctx->count[1];
    ctx->wbuf[15] = ctx->count[0];
    sha1_compile(ctx);

    /* extract the hash value as bytes in case the hash buffer is   */
    /* misaligned for 32-bit words                                  */
    for (i = 0; i < SHA1_DIGEST_SIZE; ++i)
        hval[i] = ((ctx->hash[i >> 2] >> (8 * (~i & 3))) & 0xff);
}

void sha1(unsigned char hval[], const unsigned char data[], unsigned long len)
{   
	 sha1_ctx cx[1];

    sha1_begin(cx); sha1_hash(data, len, cx); sha1_end(hval, cx);
}

#endif
